addons/survey/controllers/main.py

```diff
@@ -111,32 +105,29 @@ def start_survey(self, survey, token=None, **post):
                 type='http', auth='public', website=True)
     def fill_survey(self, survey, token, prev=None, **post):
         '''Display and validates a survey'''
-        cr, uid, context = request.cr, request.uid, request.context
-        survey_obj = request.registry['survey.survey']
-        user_input_obj = request.registry['survey.user_input']
+        SurveySurvey = request.env['survey.survey']
+        UserInput = request.env['survey.user_input']
 
         # Controls if the survey can be displayed
-        errpage = self._check_bad_cases(cr, uid, request, survey_obj, survey, user_input_obj, context=context)
+        errpage = self._check_bad_cases(request, survey)
         if errpage:
             return errpage
 
         # Load the user_input
         try:
-            user_input_id = user_input_obj.search(cr, SUPERUSER_ID, [('token', '=', token)])[0]
+            user_input = UserInput.sudo().search([('token', '=', token)], limit=1)
         except IndexError:  # Invalid token
```

Comment:
same here use if...else
to test code try with invalid/wrong token
I think no redirect to 403